### DISCUSSION TITLE
How does brew authenticate with GitHub?

### DISCUSSION THREAD

**Discussion by cdayjr** [Homebrew/discussions#1335] [Everyday usage] [answered]
I know of the `HOMEBREW_GITHUB_API_TOKEN` environment variable. I've also seen it recommend changing a repo's remote to use ssh to avoid putting in a username/password when doing `brew update`. Ae there other ways? If I authenticate with the `gh` tool, does homebrew use that?

I swear I had it working fine the other day when I didn't have `HOMEBREW_GITHUB_API_TOKEN` set up, and the only way I authenticated was via `gh`. Today it seems like even if I have `HOMEBREW_GITHUB_API_TOKEN` set to the token, it will still ask for my GitHub username and password. I could be totally wrong and something else was going on though.

Any ideas?

**Comment by carlocab** [comment, marked as answer]
I believe `brew` tries to authenticate via `git`, which should be through `credential.helper=osxkeychain` (on macOS). If you're having authentication issues it might help to [replace the credentials](https://docs.github.com/en/github/getting-started-with-github/updating-credentials-from-the-macos-keychain) stored in your keychain.

A likely explanation is that you had your GitHub password saved in your keychain, but GitHub is phasing out the use of passwords for CLI authentication which is why your stored credentials don't work.

**Reply by cdayjr** [reply]
Ah, this makes a lot of sense!

I'm logging in over ssh and I don't think the macOS credential helper is working correctly if I'm not logged in properly. I logged in and ran `gh` again and it seemed to work fine in the terminal there.

I went back to ssh and it's now printing out:
```
auth git-credential get: auth: command not found
```
- I think this might be an issue with what `gh` installs in the credential helper than anything with homebrew though. Interesting that it's missing `gh` at the start of the call.

Another question about this, does `HOMEBREW_GITHUB_API_TOKEN` not override the credential helper setting? Seems like I'm still getting issues even when I set that.

**Reply by carlocab** [reply]
Setting `HOMEBREW_GITHUB_API_TOKEN` should override the credential helper, so I'm afraid I don't know why you're still experiencing problems.

I don't think authentication with `gh` is related though. Last I checked `gh` stores credentials in `~/.config/gh/hosts.yml` (which is plain-text, sadly), which definitely isn't where `brew` and `git` look.

**Comment by tjluoma** [comment]
I have always had success by generating an ssh-key and adding my public key to my GitHub account and thought that the was the recommended way of doing it.
